### ISSUE TITLE
Week 4B: Assemble Transformer block and decoder-only LM

### ISSUE THREAD

**Issue by Liinn0529** [Liinn0529/tinygpt-from-scratch#5]
## Goal

Assemble the previously tested components into a full decoder-only language model.

## Architecture

Pre-norm block:

```text
x -> RMSNorm -> Attention -> residual
  -> RMSNorm -> SwiGLU   -> residual
```

Then:

- token embedding
- N Transformer blocks
- final RMSNorm
- tied LM head when configured

## Requirements

- config-driven dimensions
- parameter-count utility
- logits shape: `[batch, seq, vocab]`
- optional next-token cross-entropy loss when targets are supplied
- embedding/output weight tying when enabled

## Tests

- forward shape
- loss is scalar
- backward pass
- tied weights share storage/reference as intended
- parameter count remains stable for a fixed debug config
- tiny model can run on CPU

## Acceptance criteria

A debug-model forward/backward pass succeeds and all tests pass.
